From 4526c40b35d1d34c37b03363767a983f6a6ae3d6 Mon Sep 17 00:00:00 2001
From: Ryan Gonzalez <git@refi64.dev>
Date: Mon, 24 Apr 2023 23:18:57 +0000
Subject: [PATCH] gl: Try older GLES versions on EGL_BAD_ATTRIBUTE

The existing code falls back to trying older GLES versions on
EGL_BAD_MATCH, but ANGLE instead returns EGL_BAD_ATTRIBUTE if the
requested GLES version is supported by it but not the host:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/angle/src/libANGLE/validationEGL.cpp;l=2550-2561;drc=aa5b97de88cb563bb45a8a54b90bc4b3a438125d

Bug: 1400043
Change-Id: I24287668f8b10040aecd965ef38019014c0fb1d6
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4460955
Commit-Queue: Peng Huang <penghuang@chromium.org>
Reviewed-by: Peng Huang <penghuang@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1134902}
---
 ui/gl/gl_context_egl.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ui/gl/gl_context_egl.cc b/ui/gl/gl_context_egl.cc
index 08f44200745a4..74b393f81ecd6 100644
--- a/ui/gl/gl_context_egl.cc
+++ b/ui/gl/gl_context_egl.cc
@@ -352,7 +352,8 @@ bool GLContextEGL::Initialize(GLSurface* compatible_surface,
   // it might indicate that an unsupported ES version was requested. Try
   // falling back to a lower version.
   GLint error = eglGetError();
-  if (gl_display_->ext->b_EGL_KHR_no_config_context && error == EGL_BAD_MATCH) {
+  if (gl_display_->ext->b_EGL_KHR_no_config_context &&
+      (error == EGL_BAD_MATCH || error == EGL_BAD_ATTRIBUTE)) {
     // Set up the list of versions to try: 3.1 -> 3.0 -> 2.0
     std::vector<std::pair<EGLint, EGLint>> candidate_versions;
     if (context_client_major_version == 3 &&
-- 
2.39.2

